.travis.yml

```diff
@@ -1,31 +1,65 @@
+env:
+  global:
+    # include $HOME/.local/bin for `aws`
+    - PATH=$HOME/.local/bin:$PATH
 language: shell
 notifications:
   email:
     on_success: never
 stages:
-- tests
+- unit_tests
+- build
+- e2e_tests
 - deployment
+before_install:
+  # set up awscli packages
+  - pip install --user awscli
+  # set up shared storage for build artifacts
+  - mkdir -p ~/build-stages-shared-storage/$TRAVIS_BUILD_NUMBER
+  - aws s3 sync s3://aws-log-forwader-travis-build-shared-storage/$TRAVIS_BUILD_NUMBER ~/build-stages-shared-storage/$TRAVIS_BUILD_NUMBER
 jobs:
   include:
-    - stage: tests
-      name: Tests & Linting
+    - stage: unit_tests
+      name: Execute unit tests & linting
       language: shell
       services: docker
       script:
         - docker build -f pipeline/Dockerfile -t lambda_image .
         - docker run --rm lambda_image bash -c 'python3 -m pylint --rcfile=pipeline/pylint.cfg $(find src/ -name "*.py")'
         - docker run --rm lambda_image bash -c 'cd src && python3 -m pytest -v ../tests/unit'
+    - stage: build
+      name: Build AWS Lambda package
+      #if: tag =~ /^release.*$/
+      language: shell
+      services: docker
+      script:
+        - docker run --rm -v `pwd`:/var/task amazon/aws-sam-cli-build-image-python3.8 bash -c './build-release-package.sh'
+        - cp -p dynatrace-aws-log-forwarder.zip ~/build-stages-shared-storage/$TRAVIS_BUILD_NUMBER

```

Comment:
How does this work? When is stuff from ~/build-stages-shared-storage pushed to s3?

Reply:
Ok, please confirm if I understand correctly. "before_install" above which installs aws-cli and does s3-sync is executed before each step?
But still then, when is this stuff here pushed to s3? (we need some push after this step here, right?)
Also, could before_install use some docker image so we don't run install aws-cli from zero each time?

Reply:
Yes, `before_install` and `after_success` are executed with each stage. This is mainly because those steps do not take too much time and applying them to each step (`build`, `e2e_tests` and `deployment`) independently would produce repeatable code. The S3 sync is exected before each stage for the packages to be downloaded and after for the build products to be pushed.

As for the Docker image, AFAIK Travis does not allow for custom images to be used as build agents.

Reply:
Ok, I see it now! Very nice. I was confused by `after_success` since it is thrown out at the bottom and somehow I intuitively thought it is executed only once after successful build.